Mathlib/LinearAlgebra/Matrix/Ideal.lean

```diff
@@ -113,141 +114,207 @@ end Ideal
+
+end NonAssocSemiring
+
+end RingCon
+
 namespace TwoSidedIdeal
 open Matrix
 
-variable {R : Type*} [Ring R]
-         (n : Type*) [Fintype n]
+variable {R : Type*} (n : Type*)
+
+section NonUnitalNonAssocRing
+variable [NonUnitalNonAssocRing R] [Fintype n]
 
 /-- The two-sided ideal of matrices with entries in `I ≤ R`. -/
-def matricesOver (I : TwoSidedIdeal R) : TwoSidedIdeal (Matrix n n R) :=
-  TwoSidedIdeal.mk' { M | ∀ i j, M i j ∈ I }
-    (fun _ _ => I.zero_mem)
-    (fun ha hb i j => I.add_mem (ha i j) (hb i j))
-    (fun ha i j => I.neg_mem (ha i j))
-    (fun ha i j => by
-      rw [mul_apply]
-      apply sum_mem
-      intro k _
-      apply I.mul_mem_left _ _ (ha k j))
-    (fun ha i j => by
-      rw [mul_apply]
-      apply sum_mem
-      intro k _
-      apply I.mul_mem_right _ _ (ha i k))
+@[simps]
+def matricesOver (I : TwoSidedIdeal R) : TwoSidedIdeal (Matrix n n R) where
+  ringCon := I.ringCon.matrix n
 
 @[simp]
 lemma mem_matricesOver (I : TwoSidedIdeal R) (M : Matrix n n R) :
-    M ∈ I.matricesOver n ↔ ∀ i j, M i j ∈ I := by
-  simp [matricesOver]
+    M ∈ I.matricesOver n ↔ ∀ i j, M i j ∈ I := Iff.rfl
 
 theorem matricesOver_monotone : Monotone (matricesOver (R := R) n) :=
   fun _ _ IJ _ MI i j => IJ (MI i j)
 
 theorem matricesOver_strictMono_of_nonempty [h : Nonempty n] :
     StrictMono (matricesOver (R := R) n) :=
-  matricesOver_monotone n |>.strictMono_of_injective <| fun I J eq => by
-    ext x
-    have : _ ↔ _ := congr((Matrix.of fun _ _ => x) ∈ $eq)
-    simpa only [mem_matricesOver, of_apply, forall_const] using this
+  matricesOver_monotone n |>.strictMono_of_injective <|
+    .comp (fun _ _ => mk.inj) <| (RingCon.matrix_injective n).comp ringCon_injective
 
 @[simp]
-theorem matricesOver_bot : (⊥ : TwoSidedIdeal R).matricesOver n = ⊥ := by
-  ext M
-  simp only [mem_matricesOver, mem_bot]
-  constructor
-  · intro H; ext; apply H
-  · intro H; simp [H]
+theorem matricesOver_bot : (⊥ : TwoSidedIdeal R).matricesOver n = ⊥ :=
+  ringCon_injective <| RingCon.matrix_bot _
 
 @[simp]
-theorem matricesOver_top : (⊤ : TwoSidedIdeal R).matricesOver n = ⊤ := by
-  ext; simp
+theorem matricesOver_top : (⊤ : TwoSidedIdeal R).matricesOver n = ⊤ :=
+  ringCon_injective <| RingCon.matrix_top _
 
-theorem asIdeal_matricesOver [DecidableEq n] (I : TwoSidedIdeal R) :
-    asIdeal (I.matricesOver n) = (asIdeal I).matricesOver n := by
-  ext; simp
+end NonUnitalNonAssocRing
+
+section NonAssocRing
+variable [NonAssocRing R] [Fintype n] [Nonempty n] [DecidableEq n]
 
-variable {n : Type*} [Fintype n]
+variable {n}
 
 /--
 Two-sided ideals in $R$ correspond bijectively to those in $Mₙ(R)$.
 Given an ideal $I ≤ R$, we send it to $Mₙ(I)$.
 Given an ideal $J ≤ Mₙ(R)$, we send it to $\{Nᵢⱼ ∣ ∃ N ∈ J\}$.
 -/
 @[simps]
-def equivMatricesOver (i j : n) : TwoSidedIdeal R ≃ TwoSidedIdeal (Matrix n n R) where
+def equivMatricesOver [Nonempty n] [DecidableEq n] :
+    TwoSidedIdeal R ≃ TwoSidedIdeal (Matrix n n R) where
```

Comment:
Can you please add the example to the docstring explaining why it's necessarily unital?

> It's necessary. If R=2ℤ and I=4ℤ, then [[R,R],[R,I]] is a two-sided ideal in M₂(R) (in fact you're free to replace any of the R by I).

That way no one has to think about it again later.

Reply:
Are you happy with it at https://github.com/leanprover-community/mathlib4/pull/23641/files#r2032107011?